Code/Framework/AzFramework/AzFramework/Quality/QualityCVarGroup.cpp

```diff
@@ -0,0 +1,244 @@
+        return levelIndex;
+    }
+
+    QualityCVarGroup::ValueType QualityCVarGroup::GetHighestQualityForSetting(AZStd::string_view path)
+    {
+        ValueType maxQualityLevel = -1;
+        auto callback = [&]([[maybe_unused]] const AZ::SettingsRegistryInterface::VisitArgs& visitArgs)
+        {
+            maxQualityLevel++;
+            return AZ::SettingsRegistryInterface::VisitResponse::Continue;
+        };
+        AZ::SettingsRegistryVisitorUtils::VisitArray(*m_settingsRegistry, callback, path);
+        return maxQualityLevel;
+    }
+
+    void QualityCVarGroup::LoadQualityLevel(ValueType qualityLevel)
+    {
+        using FixedValueString = AZ::SettingsRegistryInterface::FixedValueString;
+        using VisitResponse = AZ::SettingsRegistryInterface::VisitResponse;
+        using Type = AZ::SettingsRegistryInterface::Type;
+
+        if (qualityLevel >= m_numQualityLevels || qualityLevel < 0)
+        {
+            AZ_Warning("QualityCVarGroup",
+                       false,
+                       "Invalid quality level %lld requested for %s which only supports quality levels 0..%lld",
+                       qualityLevel,
+                       m_functor->GetName(),
+                       m_numQualityLevels - 1);
+            return;
+        }
+
+        auto key = FixedValueString::format("%s/Settings", m_path.c_str());
+        auto settingsVisitorCallback = [&](const AZ::SettingsRegistryInterface::VisitArgs& visitArgs)
+        {
+            if (visitArgs.m_type == Type::Object ||
+                visitArgs.m_type == Type::Null ||
+                visitArgs.m_type == Type::NoType)
+            {
+                AZ_Warning("QualityCVarGroup", false, "Invalid setting value type for %.*s, objects and null are not supported, only arrays, bool, string, or numbers.", AZ_STRING_ARG(visitArgs.m_fieldName));
+                return VisitResponse::Skip;
+            }
+
+            AZStd::string_view command = visitArgs.m_fieldName;
+            AZStd::string_view valueKey = visitArgs.m_jsonKeyPath;
+            if (visitArgs.m_type == Type::Array)
+            {
+                valueKey = FixedValueString::format("%.*s/%d", AZ_STRING_ARG(valueKey), qualityLevel);
+                if (visitArgs.m_registry.GetType(valueKey) == Type::NoType)
+                {
+                    // if the array index doesn't exist, use the value for the highest index
+                    ValueType maxQualityLevel = GetHighestQualityForSetting(visitArgs.m_jsonKeyPath);
+                    if (maxQualityLevel < 0)
+                    {
+                        AZ_Warning("QualityCVarGroup", false, "Not applying settings for %.s because the array is empty.", AZ_STRING_ARG(command));
+                        return VisitResponse::Continue;
+                    }
+                    valueKey = FixedValueString::format("%.*s/%d", AZ_STRING_ARG(visitArgs.m_jsonKeyPath), maxQualityLevel);
+                }

```

Comment:
Instead of trying to guess the array index, a call to `SettingsRegistryVisitorUtils::VisitArray` can be invoked instead.

A lambda can be invoked inside of another lambda.
```suggestion
                auto GetHighestQualityLevel = [&](const AZ::SettingsRegistryInterface::VisitArgs& visitArrayArg)
                {
                         QualityLevels qualityLevel{ QualityLevels::LevelFromDeviceRules};
                         if (!visitArrayArg.m_registry.GetObject(qualityLevel, visitArrayArg.m_jsonKeyPath)
                         {
                             return VisitResponse::Continue;
                         }
                         
                         valueKey = visitArgs.m_jsonKeyPath;
                };

                AZ::SettingsRegistryVisitorUtils::VisitArray(visitArgs.m_registry, GetHighestQualityLevel, valueKey);
```

Reply:
I seem to recall I was trying to do this but VisitArray wants a non-const ref and visitArgs is a const ref.

Reply:
Ah the SettingsRegistry reference that it takes is [non-const](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Settings/SettingsRegistryVisitorUtils.h#L75-L79).

You can workaround this by passing in the Settings Registry, that you captured to the outer lambda and use that.

Alternatively we could just add const to the Visit functions themselves.

The [SettingsRegistry::Visit](https://github.com/o3de/o3de/blame/development/Code/Framework/AzCore/AzCore/Settings/SettingsRegistry.h#L228) function is already a const member function, so the Visit overloads can accept a `const SettingsRegistryInterface&`

Reply:
Now that I look at the API more, the [VisitField/VisitObject/VisitArray](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Settings/SettingsRegistryVisitorUtils.h#L70-L84) functions have no-reason why they shouldn't accept a `const SettingsRegistryInterface&`

We should update those function signatures.

Reply:
I assumed they were non-const to allow mutations to the registry

Reply:
I've gone with capturing the setting registry to have a non-const version.

Reply:
Them being non-const were a mistake, because visitation doesn't mutate the registry.

If the user wants to use an operation that mutates the registry during visitation, they can just capture a non-const reference of it in a lambda.